### DISCUSSION TITLE
Testing lifespan along with state in FastAPI

### DISCUSSION THREAD

**Discussion by Partho** [fastapi/fastapi#10800] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import httpx
from fastapi import FastAPI
from contextlib import asynccontextmanager
from typing import AsyncIterator, TypedDict

class State(TypedDict):
    async_client: httpx.AsyncClient

@asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncIterator[State]:
    async with httpx.AsyncClient() as client:
        yield {"async_client": client}

# create app
app = FastAPI(
    lifespan=lifespan,
    ...
)
```

### Description

I am trying to test lifespan function with state dict being populated. The lifespan is defined in the example code section. I am trying to use lifespan in fastapi TestClient. How will the TestClient get `async_client` from lifespan? How can we patch the state object in test? My testClient setup looks like:

```
from fastapi.testclient import TestClient

@pytest.fixture
def test_app(mocker):
    import app
    return TestClient(app=app.app)
```
I used TestClient as a context manager wherever required, to ensure that the lifespan is called.

### Operating System

macOS

### Operating System Details

MacOS Sonoma 14.2

### FastAPI Version

0.95.1

### Pydantic Version

1.10.13

### Python Version

3.11.5

**Comment by Ventura94** [comment, marked as answer]
I tried to do what you asked for and I couldn't find a way to do it either, however, I would recommend creating a dependency so that there is and closes the connections of the asynchronous httpx client:
````python
async def async_client():
    async with httpx.AsyncClient() as client:
        yield client
````
later you can overwrite your dependency in the tests:
````python
async def async_client_test():
    async with httpx.AsyncClient() as client:
        yield client


@pytest.fixture(scope="function")
def client():
    app.dependency_overrides[async_client] = async_client_test
    with TestClient(app=app) as client:        
            yield client

def test_app(client):
    ...
````
i hope it helps.
